Aerospike migrations module

Alb0t opened the pull request:
##### SUMMARY
Aerospike is a shared-nothing NoSQL database which is typically ran against 10-120 servers. With the typical replication and redundancy being able to withstand the loss of a limited amount of servers at a time, we need a module that allows us to check for replication or 'migrations' is complete to ensure data availability - before taking a system down. This ansible module should solve that need.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
aerospike_migrations module

##### ADDITIONAL INFORMATION
More info on Migrations:
https://www.aerospike.com/docs/operations/manage/migration/

<!--- Paste verbatim command output below, e.g. before and after your change -->
```$ ansible-playbook -i test.inv  test.yml -u root                                                                                                         

PLAY [upgrade aerospike] *********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [192.168.50.15]

TASK [Install dependencies] ******************************************************************************************************************************************************************************************************************
ok: [192.168.50.15]

TASK [setup aerospike] ***********************************************************************************************************************************************************************************************************************
ok: [192.168.50.15]

TASK [wait_for_migs] *************************************************************************************************************************************************************************************************************************
fatal: [192.168.50.15]: FAILED! => {"changed": false, "message": "Migrations still found after reaching limit.", "msg": "Migrations still found after reaching tries limit.", "original_message": ""}                                        
        to retry, use: --limit @/Users/alb0t/code/ansible_ansible/albot_ansible/test.retry

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
192.168.50.15              : ok=3    changed=0    unreachable=0    failed=1    skipped=0

(venv) alb0t @ ~/code/ansible_ansible/albot_ansible - [aerospike_migrations] $ ansible-playbook -i test.inv  test.yml -u root                                                                                                         

PLAY [upgrade aerospike] *********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [192.168.50.15]

TASK [Install dependencies] ******************************************************************************************************************************************************************************************************************
ok: [192.168.50.15]

TASK [setup aerospike] ***********************************************************************************************************************************************************************************************************************
ok: [192.168.50.15]

TASK [wait_for_migs] *************************************************************************************************************************************************************************************************************************
ok: [192.168.50.15]

TASK [reboot] ********************************************************************************************************************************************************************************************************************************
^C
```

Commit:
Push aerospike migrations module

Commit:
maxsplit only valid for python3

exception handling

Alb0t commented:
.

Alb0t commented:
ready_for_review

Alb0t commented:
ready_for_review

Alb0t commented:
bot_status

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+    - This can be used to check for migrations in a cluster.
+    - This makes it easy to do a rolling upgrade/update on Aerospike nodes.
+    - If waiting for migrations is not desired, simply just poll until
+    - port 3000 if available or asinfo -v status returns ok
```
This should not be a new paragraph, but continue the previous paragraph (item), right?
```suggestion
      port 3000 if available or asinfo -v status returns ok
```

Alb0t replied:
will fix.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+        default: 3000
+    connect_timeout:
+        description:
+            - How long to try to connect to node before giving up (ms)
```
Writing out `ms` as `milliseconds` might make it a little easier to understand for some people.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+    consecutive_good_checks:
+        description:
+            - How many times should the cluster report "no migrations"
+            - consecutively before returning OK back to ansible?
```
```suggestion
              consecutively before returning OK back to ansible?
```
Same as above. Every item in this list will be rendered as a new paragraph.

felixfontein replied:
(Happens several times more. I guess you will find them yourself :) )

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+    target_cluster_size:
+        description:
+            - When all aerospike builds in the cluster are greater than
+            - version 4.3, then the 'cluster-stable' info command will be used.
```
```suggestion
            - version 4.3, then the C(cluster-stable) info command will be used.
```
(see [here](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation) for some more formatting options)

Alb0t replied:
What is this notation called? never seen this
```
C(cluster-stable)

felixfontein replied:
Check the link I used above; it contains all formatting directives for module docs: https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+'''
+
+RETURN = '''
+# im not real sure what goes here
```
In case your module returns some data (like information returned by Aerospike), you can document that here. If there isn't anything, you can simply leave this empty (or keep a comment like this :) ).

Alb0t replied:
It will only return a boolean. The function doesn't change anything.. If false it should return a reason. So I guess I will notate that as a comment?

felixfontein replied:
If it's just the boolean `changed`, you can leave it empty.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+    # state will include any data that you want your module to pass back
+    # for consumption, for example, in a subsequent task
+    result = dict(
+        changed=False,
```
The contents in here are what you should document in `RETURN` above.

felixfontein replied:
(except `changed`, that doesn't need documentation)

Alb0t replied:
if it returns a reason why it failed, should it go here? i assume fail_json should be sufficient enough..

felixfontein replied:
Only put stuff here which you return in *every* run. `original_message` should not be there (you don't use that at all). `message` seems to be something you (almost) always return, so you can keep it here; but then you also have to document it in `RETURN` above.

Things used only for `fail_json()` shouldn't be here.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+    if not LIB_FOUND:
+        module.fail_json(
+            msg="A required module was not found. This playbook" +
+            " requires 'aerospike' 'time' and 're' modules. " +
```
```suggestion
            " requires the 'aerospike', 'time' and 're' modules. " +
```

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+        return result
+
+    # manipulate or modify the state as needed (this is going to be the
+    # part where your module will do what it needs to do)
```
I don't think this comment is still needed?

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+    # migrations.has_migs accepts 1 argument,
+    # a bool to specify checking local node only or entire cluster
+    has_migrations, skip_reason = migrations.has_migs(module.params['local_only'])
+    if has_migrations is False:
```
```suggestion
    if not has_migrations:
```

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,562 @@
+            msg="Failed.", **result
+        )
+
+    # use whatever logic you need to determine whether or not this module
```
I guess you could clean up some more comments.

felixfontein reviewed (commented):
I'll continue with the second part tomorrow.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+        )
+
+    if module.check_mode:
+        return result
```
You should always use `module.exit_json()` or `module.jail_json()` to return.

Also, this implementation of `check_mode` looks strange. A `check_mode` run should behave like a proper run from what it returns (if that can be determined in a reasonable way), with the exception that no change is made.

Alb0t replied:
Does the fail_json code on 215 not look right?

felixfontein replied:
That one's ok. Just using `return` is not, though :)

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+    if module.check_mode:
+        return result
+
+    migrations = Migrations(module)
```
If `Migrations()` doesn't have unwanted side-effects (like changing something), you should add `if module.check_mode:` after this. You could for example do it as follows:
```.py
if module.check_mode:
    has_migrations, skip_reason = True, None
else:
    has_migrations, skip_reason = migrations.has_migs(module.params['local_only'])
```
Then continue as before with `if not has_migrations: ...`. (I'm assuming that `True, None` is the more common return value which is to be expected.)

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+        result['message'] = "No migrations"
+    else:
+        result['message'] = skip_reason
+        module.fail_json(
```
I would change this to `module.fail_json(msg="Failed.", skip_reason=skip_reason)`, so that `skip_reason` is returned explicitly.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+    # a bool to specify checking local node only or entire cluster
+    has_migrations, skip_reason = migrations.has_migs(module.params['local_only'])
+    if not has_migrations:
+        result['message'] = "No migrations"
```
I don't think you need to return this fixed string on success. Just not returning anything should be fine.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+    # TODO: add support for auth, tls, and other special features
+    # I won't use those features, so I'll wait until somebody complains
+    # or does it for me (Cross fingers)
+    # create the client object
```
Python uses [docstrings](https://en.wikipedia.org/wiki/Docstring#Python) for method documentation, not comments.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+        # so we dont want a dict from those.
+        if '=' in data:
+            retval = dict(
+                metric.split("=") for metric in data.split(delimiter))
```
What happens if you have something like `a=b=c`? Then you will init the dictionary with the tuple `(a, b, c)`, which will fail.

Usually you want to interpret `a=b=c` as key `a` with value `b=c`; for that, you need `metric.split("=", maxsplit=1)`.

felixfontein replied:
Also, why do you compute `data.split(delimiter)` again instead of using `data_arr`?

```suggestion
                metric.split("=", maxsplit=1) for metric in data_arr)
```

Alb0t replied:
This is actually a scenario for some of the already exposed metrics. They aren't used by this module, but very good feedback. I will add it! Thanks for the tip, didn't know about this.

Alb0t replied:
Probably a brain fart on the data_arr lol

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+                str(len(data))
+            )
+        # data will be in format 'command\touput'
+        data = data[1]
```
This will fail if `len(data) == 1`, which you explicitly allow above.

Alb0t replied:
Not base 0, aah!

Alb0t replied:
Why will this fail with len(data)==1?

`
        data = self._client.info_node(cmd, node)
        data = data.split("\t")
        if len(data) != 1 and len(data) != 2:
            self.module.fail_json(
                msg="Unexpected number of values returned in info command: " +
                str(len(data))
            )
        # data will be in format 'command\touput'
        data = data[1]
`

1. data="commandoutput\t
x=2;y=3;x=bat;foo=bar;etc=etc;etc=etc;"
2. data=[commandoutput, x=2;y=3;x=bat;foo=bar;etc=etc;etc=etc;]
3. len(data) is 2... 
oh but ill always be accessing [1] which is the second element. got it!
I think I will use [-1] here instead.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+        # if version <4.3 we can't use cluster-stable info cmd
+        # regex hack to check for versions beginning with 0-3 or
+        # beginning with 4.0,4.1,4.2
+        if re.search('^([0-3]|4.[0-2])', min(self._build_list)):
```
This would also match version 10.x. Also, `4.[0-2]` will also match `421` (an unescaped dot matches anything).

You probably want:
```suggestion
        if re.search('^([0-3]\\.|4\\.[0-2])', min(self._build_list)):
```
Or (using raw strings, to avoid double escaping):
```suggestion
        if re.search(R'^([0-3]\.|4\.[0-2])', min(self._build_list)):
```

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+                int(namespace_stats[self.module.params['migrate_tx_key']])
+            namespace_rx = \
+                int(namespace_stats[self.module.params['migrate_tx_key']])
+            test = int(namespace_tx) + 1
```
I think you can simply delete this and the next line. `namespace_tx` and `namespace_rx` are already `int`s at this point.

Alb0t replied:
I wanted to make sure it was type checked. I guess I'm just being paranoid.. will kill it :)

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+            self.module.fail_json(
+                msg="namespace stat returned was not numerical"
+            )
+        if namespace_tx != 0 or namespace_rx != 0:
```
Why don't you replace lines 365 to 371 with `return namespace_tx != 0 or namespace_rx != 0`?

Alb0t replied:
terse code scares me, but it's not horribly terse so i'll take it

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+        for namespace in self._namespaces:
+            if self._namespace_has_migs(namespace, node) is True:
+                migs += 1
+        if migs == 0:
```
`return migs != 0` is equivalent ot lines 380 to 382.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+            return True
+        return False
+
+    '''Added 4.3:
```
If this docstring is for the `_cluster_stable()` method, it must be *inside* that method.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+                            "Skipping on try#" + str(try_num) +
+                            " for reason:" + " cluster_stable"
+                        )
+                elif self._has_migs(local) is True:
```
Use `==` to compare booleans. The `is` test should *only* be used when you really want to compare identities of objects!

Finally, there's no need to compare a boolean to `True` or `False` explicitly; simply write `elif self._has_migs(local):`.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+                        " for reason:" + " migrations"
+                    )
+                    consecutive_good = 0
+                elif self._has_migs(local) is False:
```
Don't you simply want an `else:` here? Or do you explicitly want to have the case that you call `self._has_migs(local)` twice in a row (without a sleep inbetween), and once it returns `False` and then `True`? Otherwise this `elif` makes no sense.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,535 @@
+
+def main():
+    """main method for ansible module"""
+    run_module()
```
Why don't you put the contents of `run_module()` into `main()`? Is there an advantage of having this additional indirection?

Alb0t replied:
Just was following the template, wasn't sure if there was a dependency on that method being there. Think I should move it?

felixfontein replied:
Ah, I didn't knew there was a template :) Well, in that case, keep it. It doesn't really hurt.

felixfontein commented:
Also, you might want to reduce your existing commit count (by squashing some of them together). Especially when you rebase anyway :)

Alb0t commented:
welp, i failed at squash.

Alb0t commented:
ready_for_review

felixfontein commented:
> welp, i failed at squash.

The good thing is, you can try again the next time ;)

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,523 @@
+    if module.check_mode:
+        has_migrations, skip_reason = True, None
+    else:
+        has_migrations, skip_reason = migrations.has_migs(
```
Don't you need `migrations` (from three lines below) here? Otherwise you cannot use that object.

Alb0t replied:
Maybe I'm not seeing what you're talking about... or i changed it already :S.. where am i calling it before it is defined?

felixfontein replied:
I don't know whether you changed it when you wrote this, but now it's correct ;-)

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,523 @@
+                    consecutive_good = 0
+                else:
+                    consecutive_good += 1
+                    if consecutive_good is self.module.params[
```
https://stackoverflow.com/a/2239753
```suggestion
                    if consecutive_good == self.module.params[
```

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,523 @@
+    result = dict(
+        changed=False,
+        original_message='',
+        message=''
```
None of `message` and `original_message` are actually used by your module (and they are not documented in `RETURN` either). Please just remove them if you don't want to change that.

Alb0t replied:
ok, wasn't sure if it would freak out if i left it undefined.

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,523 @@
+        )
+
+    if module.check_mode:
+        has_migrations, skip_reason = True, None
```
So the common case is that `has_migrations` is `True`, i.e. the module fails?

Alb0t replied:
in most cases it would return False, so ill switch it over

Alb0t commented:
Well I rebased to devel, did rebase HEAD~20 and force pushed. thought that was all that was needed. im horrible at this..

Alb0t commented:
Ah, need to replace 'pick' with 'squash'. nifty. Not quite the commit message I wanted but I think it works

Alb0t commented:
ready_for_review

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,520 @@
+    )
+    if not LIB_FOUND:
+        module.fail_json(
+            msg="A required module was not found. This playbook" +
```
```suggestion
            msg="A required Python library was not found. This module" +
```
That's more precise :) Also, you can assume that the `time` and `re` modules are installed; if not, a lot of other modules will also break and the user has a really broken Python installation, which should happen so seldomly that an extra error message for that should be unnecessary.

felixfontein reviewed (approved):
From my side it looks good. I don't have any experience with Aerospike, though, so I have no idea whether this actually works -- I assume you've tested it :)

Alb0t commented:
Doing another round of testing, got this:
```
fatal: [192.168.50.124]: FAILED! => {"attempts": 2, "changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"<stdin>\", line 113, in <module>\n  File \"<stdin>\", line 105, in _ansiballz_main\n  File \"<stdin>\", line 48, in invoke_module\n  File \"/tmp/ansible_aerospike_migrations_payload_36kHd2/__main__.py\", line 520, in <module>\n  File \"/tmp/ansible_aerospike_migrations_payload_36kHd2/__main__.py\", line 516, in main\n  File \"/tmp/ansible_aerospike_migrations_payload_36kHd2/__main__.py\", line 222, in run_module\n  File \"/tmp/ansible_aerospike_migrations_payload_36kHd2/__main__.py\", line 242, in __init__\n  File \"/tmp/ansible_aerospike_migrations_payload_36kHd2/__main__.py\", line 328, in _update_cluster_statistics\n  File \"/tmp/ansible_aerospike_migrations_payload_36kHd2/__main__.py\", line 286, in _info_cmd_helper\n  File \"/tmp/ansible_aerospike_migrations_payload_36kHd2/__main__.py\", line 286, in <genexpr>\nTypeError: split() takes no keyword arguments\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

Stackoverflow says the maxsplit is a python3 only thing.. So I modified it based off https://stackoverflow.com/questions/27493703/typeerror-split-takes-no-keyword-arguments-in-python-2-x to just do `metric.split("=", 1) for metric in data_arr` instead of `metric.split("=", maxsplit=1) for metric in data_arr`. Any feelings on that?

Alb0t commented:
ready_for_review

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,520 @@
+        # so we dont want a dict from those.
+        if '=' in data:
+            retval = dict(
+                metric.split("=", 1) for metric in data_arr
```
Sorry about `maxsplit`, I'm mostly using Python 3 nowadays and forgot it wasn't a keyword in Python 2. Anyway, `metric.split("=", 1)` will work fine in both Python 2 and 3, so it's good!

Alb0t replied:
great :)

felixfontein commented on lib/ansible/modules/database/aerospike/aerospike_migrations.py:
```diff
@@ -0,0 +1,520 @@
+    if module.check_mode:
+        has_migrations, skip_reason = False, None
+    else:
+        migrations = Migrations(module)
```
How about adding a `try/except` around these two lines, so that in case something like the `ServerError` is raised (like in your test output), the user will get a nicer error?

```.py
try:
    migrations = Migrations(module)
    has_migrations, skip_reason = migrations.has_migs(
         module.params['local_only']
    )
except Exception as e:
    module.fail_json(msg="Error: {0}".format(e))
```

Alb0t commented:
Oh god I can't get my own module running... uhhh...
Any idea what im missing here? Tried installing aerospike with every pip binary there is... even tried specifying python path and made sure that 'import aerospike' works when running python3 shell. I'm dumbfounded...

`TASK [Check and or wait for Migrations] *******************************************************************************
task path: /Users/albert.autin/code/albert.autin/ansible/cit_coldstorage_backup.yml:21
FAILED - RETRYING: Check and or wait for Migrations (1 retries left).
FAILED - RETRYING: Check and or wait for Migrations (1 retries left).
FAILED - RETRYING: Check and or wait for Migrations (1 retries left).
fatal: [CIT_COLDS_POC1]: FAILED! => {"attempts": 1, "changed": false, "msg": "A required module was not found. This playbook requires the 'aerospike' 'time' and 're' modules. Please run 'pip install aerospike'. The other modules should be included in a basic python install. OS Error: No module named 'aerospike'"}
fatal: [CIT_COLDS_POC3]: FAILED! => {"attempts": 1, "changed": false, "msg": "A required module was not found. This playbook requires the 'aerospike' 'time' and 're' modules. Please run 'pip install aerospike'. The other modules should be included in a basic python install. OS Error: No module named 'aerospike'"}
fatal: [CIT_COLDS_POC4]: FAILED! => {"attempts": 1, "changed": false, "msg": "A required module was not found. This playbook requires the 'aerospike' 'time' and 're' modules. Please run 'pip install aerospike'. The other modules should be included in a basic python install. OS Error: No module named 'aerospike'"}`

Alb0t commented:
ready_for_review

felixfontein reviewed (approved):
shipit

(Like I said before, I'm not an aerospike user, so I didn't test this, but the code looks good. Assuming that @Alb0t did test it, it's fine to merge this for me.)

Alb0t commented:
thanks @felixfontein for all the help! We're using this actively in production with AWX. It's beautiful

Alb0t commented:
bot_status

felixfontein commented:
You need at least one more review before this can get merged. You could for example bring your PR up in a Public Ansible Meeting on IRC (Freenode); see https://github.com/ansible/community/issues/418 for details. @gundalow might also have some ideas :)
